### PR TITLE
Change PV naming and fix pointer bug

### LINKED ISSUE #3982: Porch: PackageVariantSet deletes and recreates PackageVariants instead of updating
### Expected behavior
When a change is made to a PackageVariantSet that implies changes to the fields of a PackageVariant, but that does not imply a change in the repo or package name, the expectation is that the PVS will update the existing, previously created PackageVariant.

### Actual behavior
The PVS deletes the PV, and creates a new one, with a new name. This seems to be at least one cause of #3980, though I suspect not the only cause.

This also causes existing published versions to be flagged as DeletionProposed.

### Information
When an update is made to a PV, it will create a new Draft, based on the existing, most recent *published* version of the package, and incorporate the changes to the PV in that new Draft. If we instead delete and recreate, we lose any changes the user may have made to the Draft that were later published. So, delete & recreate can be quite destructive.

### Steps to reproduce the behavior
See #3980.

### LINKED ISSUE #3985: Porch: PackageVariantSet target.name expressions do not work as expected
### Expected behavior
When using target.name in an expression, the resulting PackageVariants end up with the value corresponding to that target.name.

### Actual behavior
When using target.name in an expression, all resulting PackageVariants end up with the same value.

### Information
@henderiw found this during Nephio testing.

### PR DESCRIPTION
**Pull request by johnbelamaric**
Fixes #3982 
Fixes #3985 

And adds unit tests covering those.

### CODE REVIEW

**Comment by johnbelamaric** [comment]
@natasha41575 @mortent PTAL

File: porch/controllers/packagevariantsets/pkg/controllers/packagevariantset/packagevariantset_controller.go [hunk 1]
```diff
@@ -315,14 +312,16 @@ func (r *PackageVariantSetReconciler) ensurePackageVariants(ctx context.Context,
...
 		}
-		desiredPackageVariantMap[hash] = &pv
+		desiredPackageVariantMap[pvId] = &pv
 	}
 
-	for existingPvHash, existingPV := range existingPackageVariantMap {
-		if _, found := desiredPackageVariantMap[existingPvHash]; found {
+	for existingPvId, existingPV := range existingPackageVariantMap {
+		if _, found := desiredPackageVariantMap[existingPvId]; found {
+			fmt.Println("Found existing in desired", existingPvId)
```
Comment (mortent) [inline porch/controllers/packagevariantsets/pkg/controllers/packagevariantset/packagevariantset_controller.go:320]: Nit: Use klog for logging.

**Review by mortent — APPROVED** [review]

File: porch/controllers/packagevariantsets/pkg/controllers/packagevariantset/packagevariantset_controller.go (hunk 1 above)
Comment (johnbelamaric) [reply porch/controllers/packagevariantsets/pkg/controllers/packagevariantset/packagevariantset_controller.go:320]: Oops, that's debug, I'll remove it

**Comment by henderiw** [comment]
I tested it a few times now and seems to do the job

**Pull request merged**